clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerInterceptors.java

```diff
@@ -107,14 +110,21 @@ public void onAcknowledgement(RecordMetadata metadata, Exception exception) {
     public void onSendError(ProducerRecord<K, V> record, TopicPartition interceptTopicPartition, Exception exception) {
         for (ProducerInterceptor<K, V> interceptor : this.interceptors) {
             try {
+                Headers headers = record != null ? record.headers() : null;
+                if (headers instanceof RecordHeaders) {
+                    // make a copy of the headers to make sure it is read-only
+                    RecordHeaders recordHeaders = (RecordHeaders) headers;
+                    headers = new RecordHeaders(recordHeaders);
+                    ((RecordHeaders) headers).setReadOnly();
+                }
                 if (record == null && interceptTopicPartition == null) {
-                    interceptor.onAcknowledgement(null, exception);
+                    interceptor.onAcknowledgement(null, exception, headers);
```

Comment:
We must call `onAcknowledgement(RecordMetadata, Exception, Headers)` even if headers is null. Otherwise, users who implement only the new method will miss this event.